modules/internal/ArrayViewRankChange.chpl

```diff
@@ -22,9 +22,337 @@
+
+    // the lower-dimensional index set that we represent upwards
+    var updom: DefaultRectangularDom(rank, idxType, stridable);
+
+    // the collapsed dimensions and indices in those dimensions
+    const collapsedDim;
+    const idx;
+
+    const dist;  // a reference back to our ArrayViewRankChangeDist
+
+    // the higher-dimensional domain that we're equivalent to
+    var downdomPid:int;
+    var downdomInst: downdomtype(downrank, idxType, stridable);
+
+    //
+    // TODO: If we put this expression into the variable declaration
+    // above, we get a memory leak.  File a future against this?
+    //
+    proc downdomtype(param rank: int, type idxType, param stridable: bool) type {
+      var a = dist.downdist.newRectangularDom(rank=rank, idxType=idxType,
+                                              stridable=stridable);
+      return a.type;
+    }
+
+    proc downrank param {
+      return collapsedDim.size;
+    }
+
+    inline proc downdom {
+      if _isPrivatized(downdomInst) then
+        return chpl_getPrivatizedCopy(downdomInst.type, downdomPid);
+      else
+        return downdomInst;
+    }
+
+    proc dsiBuildArray(type eltType) {
+      pragma "no auto destroy"
+      const downarr = _newArray(downdom.dsiBuildArray(eltType));
+      return new ArrayViewRankChangeArr(eltType  =eltType,
+      // TODO: Update once we start privatizing vvv
+                                        _DomPid = nullPid,
+                                        dom = this,
+                                        _ArrPid=downarr._pid,
+                                        _ArrInstance=downarr._instance,
+                                        collapsedDim=collapsedDim,
+                                        idx=idx,
+                                        ownsArrInstance=true);
+    }
+
+    // TODO: Use delegation feature for these?
+    // TODO: Can't all these be implemented in ChapelArray given dsiDim()?
+
+    proc dsiNumIndices {
+      return updom.dsiNumIndices;
+    }
+
+    proc dsiLow {
+      return updom.dsiLow;
+    }
+
+    proc dsiHigh {
+      return updom.dsiHigh;
+    }
+
+    proc dsiAlignedLow {
+      return updom.dsiAlignedLow;
+    }
+
+    proc dsiAlignedHigh {
+      return updom.dsiAlignedHigh;
+    }
+
+    proc dsiStride {
+      return updom.dsiStride;
+    }
+
+    proc dsiAlignment {
+      return updom.dsiAlignment;
+    }
+
+    proc dsiFirst {
+      return updom.dsiFirst;
+    }
+
+    proc dsiLast {
+      return updom.dsiLast;
+    }
+
+    proc dsiDim(upDim: int) {
+      var downDim = 0;
+      for d in 1..downrank {
+        if !collapsedDim(d) {
+          downDim += 1;
+          if (upDim == downDim) then
+            return downdom.dsiDim(d);
+        }
+      }
+      halt("Fell out of loop in dsiDim():", (upDim, downDim));
+    }
+
+    proc dsiDims() {
+      var upDims: rank*range(idxType, stridable=stridable);
+      var upDim = 1;
+      for param downDim in 1..downrank {
```

Comment:
No strong reason I think.  I tend to use 'param' loops for ones over params that are expected to be small (like the # of dimensions in a domain/array), but apparently not consistently.

I think I can make this even simpler, though.  Re-reviewing this last night while working on reindexing, I found myself wondering why I didn't just pass the dsiDIm*() calls to updom where, arguably, all this work has already been done.  My best guess is that I wrote this routine before introducing updom and never thought to reconsider it.